TACC/Core-Portal
# Task/FP-547: Relaunch jobs in Job History modal.

## Overview: ##
Tried to replicate CEPv1's job relaunch behavior.

It seems like resubmit has only one api:
* In the Tapis documentation:
https://tapis-cli.readthedocs.io/en/latest/usage/jobs.html#jobs-resubmit
* Based on code from CEPv1:
https://bitbucket.org/taccaci/core-exp-portal/src/b633800261b4ef859a4bc36dacd5f8e5c0bc6897/server/portal/apps/workspace/api/views.py#lines-258
https://bitbucket.org/taccaci/core-exp-portal/src/b633800261b4ef859a4bc36dacd5f8e5c0bc6897/client/src/angular/workspace/services/jobs-service.js#lines-71

Based on CEPv1, we are assuming that job resubmission is done when job is finished.
* Based on code from CEPv1:
https://bitbucket.org/taccaci/core-exp-portal/src/b633800261b4ef859a4bc36dacd5f8e5c0bc6897/client/src/angular/jobs/components/job-actions/job-actions.template.html#lines-4

Design of the relaunch button was based on:
- https://xd.adobe.com/view/db2660cc-1011-4f26-5d31-019ce87c1fe8-ad17/screen/7d3e671f-bb61-4a20-89c6-9ca9a78a6f93/

## Related Jira tickets: ##

* [FP-547](https://jira.tacc.utexas.edu/browse/FP-547)

## Summary of Changes: ##
1. Added backend code the handle job relaunch by adding job objects as well as sending the job request to tapis.
2. Added the `Relaunch Job` button in the Job History modal.
3. Added chaining in the `SUBMIT_JOB` saga to call `GET_JOBS` after the resubmit job request.

## Testing Steps: ##
1. Submit a job that completes.
2. Once it's finished, ensure that you see the `Relaunch Job` button the Job's `View Details` modal (`workbench/history/jobs`).
3. Click `Relaunch Job`.
4. Ensure that a new task with the same parameters is launched and the previous job is removed from the list.
5. Ensure that the new task is successful.

## UI Photos:
![Screenshot from 2021-11-16 16-00-44](https://user-images.githubusercontent.com/9425579/142072938-45c8bcaf-ee9c-4615-a5c8-ac8c963ee336.png)

## Notes: ##
Questions:
1. Should we redesign it so that clicking on `Relaunch Job` opens another form that let's the user modify parameters based on the previous job submission?
2. Would it be correct to show the `Relaunch Job` button when job status is failed as well as finished?
3. Should we implement `cancelJob()` when job status is not finished? (code was already there in the backend)
4. I currently remove the previous job on clicking `Relaunch Job` and create a new job in it's place. Is this acceptable?
5. Should there be a separate `RESUBMITTING` job status?

TODO
1. [ ] Add test for dispatch events in job history modal.
